crates/telio-network-monitors/src/mac.rs

```diff
@@ -0,0 +1,103 @@
+};
+use parking_lot::Mutex;
+use std::{
+    cell::RefCell,
+    ffi::{c_long, c_void, CStr},
+    rc::Rc,
+};
+use telio_utils::{telio_log_info, telio_log_warn};
+
+use crate::monitor::PATH_CHANGE_BROADCAST;
+
+/// Dispatch queue priority as high priority queue
+pub const DISPATCH_QUEUE_PRIORITY_HIGH: c_long = 2;
+/// Dispatch queue priority as default queue
+pub const DISPATCH_QUEUE_PRIORITY_DEFAULT: c_long = 0;
+/// Dispatch queue priority as low priority queue
+pub const DISPATCH_QUEUE_PRIORITY_LOW: c_long = -2;
+/// Dispatch queue priority as background queue
+pub const DISPATCH_QUEUE_PRIORITY_BACKGROUND: c_long = -1 << 15;
+
+/// Vector containing interface names in OS order
+pub static INTERFACE_NAMES_IN_OS_PREFERENCE_ORDER: Mutex<Vec<String>> = Mutex::new(Vec::new());
+
+extern "C" {
+    /// Obj-c signature:
+    /// void nw_path_monitor_set_update_handler(nw_path_monitor_t monitor, nw_path_monitor_update_handler_t update_handler);
+    /// typedef void (^nw_path_monitor_update_handler_t)(nw_path_t path);
+    pub fn nw_path_monitor_set_update_handler(
+        monitor: nw_path_monitor_t,
+        update_handler: *const c_void,
+    );
+    /// Obj-c signature:
+    /// void nw_path_enumerate_interfaces(nw_path_t path, nw_path_enumerate_interfaces_block_t enumerate_block);
+    /// typedef bool (^nw_path_enumerate_interfaces_block_t)(nw_interface_t interface);
+    pub fn nw_path_enumerate_interfaces(path: nw_path_t, enumerate_block: *const c_void);
+}
+
+/// Setup network path monitor in Apple framework
+pub fn setup_network_monitor() {
+    let update_handler = block::ConcreteBlock::new(|path: nw_path_t| {
+        let names = Rc::new(RefCell::new(vec![]));
+        let names_copy = names.clone();
+
+        let enumerate_callback =
+            block::ConcreteBlock::new(move |interface: nw_interface_t| -> bool {
+                let c_name = unsafe { nw_interface_get_name(interface) };
+                if !c_name.is_null() {
+                    let name = unsafe { CStr::from_ptr(c_name) };
+                    if let Ok(name) = name.to_str() {
+                        names_copy.borrow_mut().push(name.to_owned());
+                    }
+                }
+                true
+            })
+            .copy();
+
+        unsafe {
+            nw_path_enumerate_interfaces(
+                path,
+                &*enumerate_callback as *const block::Block<_, _> as *const c_void,
+            )
+        };
+
+        *INTERFACE_NAMES_IN_OS_PREFERENCE_ORDER.lock() = names.borrow().clone();
+        if let Err(e) = PATH_CHANGE_BROADCAST.send(()) {
+            telio_log_warn!(
+                "Failed to notify about changed path, error: {e}, path: {:?}",
+                names.borrow()
+            );
+        }
+
+        telio_log_info!(
+            "Path change notification sent, current network path in os preference order: {:?}",
+            names.borrow()
+        );
+    })
+    .copy();
```

Comment:
Why are we making a copy already when declaring it?

Reply:
Ok to answer to myself, this is copying the stack allocated `ConcreteBlock` to the heap as an `RcBlock` before
passing it to Objective-C, in order to prevent a double free.